src/core/app.lua

```diff
@@ -229,27 +216,79 @@ function apply_config_actions (actions, conf)
       if not new_app_table[ta] then error("no such app: " .. ta) end
       -- Create or reuse a link and assign/update receiving app index
       local link = link_table[linkspec] or link.new(linkspec)
-      link.receiving_app = app_name_to_index[ta]
       -- Add link to apps
       new_app_table[fa].output[fl] = link
       table.insert(new_app_table[fa].output, link)
       new_app_table[ta].input[tl] = link
       table.insert(new_app_table[ta].input, link)
       -- Remember link
       new_link_table[linkspec] = link
-      table.insert(new_link_array, link)
    end
    -- Free obsolete links.
    for linkspec, r in pairs(link_table) do
       if not new_link_table[linkspec] then link.free(r, linkspec) end
    end
    -- Commit changes.
    app_table, link_table = new_app_table, new_link_table
-   app_array, link_array = new_app_array, new_link_array
    -- Trigger link event for each app.
-   for _, app in ipairs(app_array) do
+   for name, app in pairs(app_table) do
       if app.link then app:link() end
    end
+   compute_breathe_order ()
+end
+
+-- Sort the NODES topologically according to SUCCESSORS via
+-- reverse-post-order numbering.  The sort starts with ENTRIES.  This
+-- implementation is recursive; we should change it to be iterative
+-- instead.
```

Comment:
I am slightly concerned that a 100-app limit could make the NFV application explode with larger configurations. Our OpenStack test suite does have a "many VMs" test cases but that exercises less than the maximum number of apps. I suppose that ideally I would either beef that test up (we have throttled it due to resource constraints in the nested-virtualized test env) or write a new test case.

Generally question on my mind is: How much stack space is okay with LuaJIT? and how much is this algorithm going to need? You seem to be plugging in values from gut feeling here? I feel like some languages blow up with around 100 stack frames while others handle millions gracefully.

I did a quick test with a recursive factorial function and it seems okay up to around 30,000 stack frames:

```
$ ./snabb snsh -e 'function factorial (n) return (n==0) and 1 or n*factorial(n-1) end  print(factorial(30000))'
inf
```

and then exhausts the stack with 35,000:

```
$ ./snabb snsh -e 'function factorial (n) return (n==0) and 1 or n*factorial(n-1) end  print(factorial(35000))'
[string "function factorial (n) return (n==0) and 1 or..."]:1: stack overflow
```
If this topological sort is using O(n) stack space then we should be fine to well beyond 100 apps, right?